xdsl/transforms/common_subexpression_elimination.py

```diff
@@ -0,0 +1,197 @@
+            and len(self.op.results) == len(other.op.results)
+            and all(r.type == o.type for r, o in zip(self.op.results, other.op.results))
+            and len(self.op.regions) == len(other.op.regions)
+            and all(
+                s.is_structurally_equivalent(o)
+                for s, o in zip(self.op.regions, other.op.regions)
+            )
+        )
+
+
+_D = TypeVar("_D")
+
+
+class KnownOps:
+    """
+    Cache dictionary for known operations used in CSE.
+    It quacks like a dict[Operation, Operation], but uses OperationInfo of an Opetration
+    as the actual key.
+    """
+
+    _known_ops: dict[OperationInfo, Operation]
+
+    def __init__(self, known_ops: "KnownOps | None" = None):
+        if known_ops is None:
+            self._known_ops = {}
+        else:
+            self._known_ops = dict(known_ops._known_ops)
+
+    def __getitem__(self, k: Operation):
+        return self._known_ops[OperationInfo(k)]
+
+    def __setitem__(self, k: Operation, v: Operation):
+        self._known_ops[OperationInfo(k)] = v
+
+    def __contains__(self, k: Operation):
+        return OperationInfo(k) in self._known_ops
+
+    def get(self, k: Operation, default: _D = None) -> Operation | _D:
+        return self._known_ops.get(OperationInfo(k), default)
+
+    def pop(self, k: Operation):
+        return self._known_ops.pop(OperationInfo(k))
+
+
+@dataclass
+class CSEDriver:
+    """
+    Boilerplate class to handle and carry the state for CSE.
+    """
+
+    rewriter: Rewriter
+    to_erase: set[Operation] = field(default_factory=set)
+    known_ops: KnownOps = KnownOps()
+
+    def simplify_operation(self, op: Operation):
+        """
+        Simplify a single operation: replace it by a corresponding known operation in
+        scope, if any.
+        Also just delete dead operations.
+        """
+        # Don't simplify terminators.
+        if op.has_trait(IsTerminator):
+            return
+
+        # If the operation is already trivially dead just add it to the erase list.
+        if is_trivially_dead(op):
+            self.to_erase.add(op)
+            return
+
+        # Don't simplify operations with regions that have multiple blocks.
+        # MLIR doesn't either at time of writing :)
+        if any(len(region.blocks) > 1 for region in op.regions):
+            return
+
+        # Here, MLIR says something like "if the operation has side effects"
+        # Using more generics analytics; and has fancier analysis for that case,
+        # where it might simplify some side-effecting operations still.
+        # Doesmn't mean we can't just simplify what we can with our simpler model :)
+        if not op.has_trait(Pure):
+            return
+
+        # This operation rings a bell!
+        if existing := self.known_ops.get(op):
+
+            # Just replace results
+            def wasVisited(use: Use):
+                return self.known_ops.get(use.operation) is None
```

Comment:
Seems like making `use.operation in self.known_ops` work would be preferable, hopefully not too complicated

Reply:
Not too complicated, as it already works :smile: Just didn't use it